cirq/circuits/circuit_dag.py

```diff
@@ -0,0 +1,122 @@
+# Copyright 2018 The ops Developers
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from typing import Any, Generic, Iterator, Type, TypeVar
+
+import networkx
+
+from cirq import ops, devices, abc
+from cirq.circuits import circuit
+
+
+T = TypeVar('T')
+
+class Unique(Generic[T]):
+    def __init__(self, val: T) -> None:
+        self.val = val
+
+    def __repr__(self):
+        return 'Unique({}, {!r})'.format(id(self), self.val)
+
+
+TSelf = TypeVar('TSelf', bound='CircuitDagAbc')
+
+class CircuitDagAbc(networkx.DiGraph, metaclass=abc.ABCMeta):
+    """A representation of a Circuit as a directed acyclic graph.  Subclass
+    CircuitDagAbc and implement can_reorder()."""
+
+    def __init__(self,
+                 incoming_graph_data: Any = None,
+                 device: devices.Device = devices.UnconstrainedDevice
+                 ) -> None:
+        super().__init__(incoming_graph_data)
+        self.device = device
+
+    @abc.abstractmethod
```

Comment:
I think, given that this is a "data structure class", it makes more sense to tell it how to do things by providing constructor arguments rather than via inheritance. For example, the way you use `defaultdict` is not by subclassing it and implementing a `make_value` method; you give the constructor a `make_value` lambda. Do that.

Reply:
Done.